libeos-updater-util/avahi-service-file.c

```diff
@@ -257,3 +326,426 @@ eos_avahi_service_file_delete (const gchar   *avahi_service_directory,
+                                        NULL);
+  return g_file_new_for_path (service_file_path);
+}
+
+static gboolean
+get_summary_timestamp (OstreeRepo *repo,
+                       GCancellable *cancellable,
+                       guint64 *out_timestamp,
+                       GError **error)
+{
+  // TODO: Find the summary without using eos specific bits to ease
+  // the moving of this code to ostree.
+  g_autoptr(EosExtensions) extensions = eos_extensions_new_from_repo (repo,
+                                                                      cancellable,
+                                                                      error);
+
+  if (extensions == NULL)
+    return FALSE;
+
+  // TODO: Try getting the timestamp from summary metadata first.
+  g_assert (out_timestamp);
+  *out_timestamp = extensions->summary_modification_time_unix;
+  return TRUE;
+}
+
+static guint16
+get_repository_index (GVariantDict *options_dict)
+{
+  guint16 index = 0;
+
+  g_variant_dict_lookup (options_dict, EOS_OSTREE_AVAHI_OPTION_REPO_INDEX_Q,
+                         "q", &index);
+
+  return index;
+}
+
+static guint16
+get_avahi_service_port (GVariantDict *options_dict)
+{
+  // TODO: Should we store the port number in the configuration
+  // instead having it as the compile-time constant? I can't remember
+  // why I opted for this solution (was it lack of the configuration
+  // file?)
+  //
+  // In case when config file doesn't specify the port number and the
+  // number wasn't provided with the options_dict variant, likely bail
+  // out.
+  guint16 port = EOS_AVAHI_PORT;
+
+  g_variant_dict_lookup (options_dict, EOS_OSTREE_AVAHI_OPTION_PORT_Q,
+                         "q", &port);
+
+  return port;
+}
+
+static gboolean
+fill_filter_with_refs (OstreeBloom   *filter,
+                       OstreeRepo    *repo,
+                       GCancellable  *cancellable,
+                       GError       **error)
+{
+  g_autoptr(GHashTable) refs = NULL;
+  GHashTableIter iter;
+  gpointer key;
+
+  if (!ostree_repo_list_refs (repo,
+                              NULL, /* no prefix filtering */
+                              &refs,
+                              cancellable,
+                              error))
+    return FALSE;
+
+  g_hash_table_iter_init (&iter, refs);
+  while (g_hash_table_iter_next (&iter, &key, NULL))
+    {
+      ostree_bloom_add_element (filter, key);
+    }
+
+  return TRUE;
+}
+
+#define variant_sink_new(format_string, ...) (g_variant_ref_sink (g_variant_new (format_string, __VA_ARGS__)))
+
+static GVariant *
+get_bloom_filter_variant (OstreeRepo *repo,
+                          GVariantDict *options_dict,
+                          GCancellable *cancellable,
+                          GError **error)
+{
+  guint32 bloom_size = 255 - strlen(EOS_OSTREE_AVAHI_V1_REFS_BLOOM_FILTER_FIELD) - 2;
+  guint8 bloom_k = 1;
+  guint8 bloom_hash_id = EOS_OSTREE_AVAHI_STR_BLOOM_HASH_ID;
```

Comment:
It would probably also be useful to add a comment identifying which parameters need to be kept stable otherwise binary compatibility of the bloom filter between server and client will be broken. (`bloom_size`, `bloom_k` and `bloom_hash_id`.)